v4/alto-4-3.xsd

```diff
@@ -0,0 +1,1223 @@
+			<xsd:annotation>
+				<xsd:documentation>Content of the substitution.</xsd:documentation>
+			</xsd:annotation>
+		</xsd:attribute>
+		<xsd:attribute name="WC" type="WCType" use="optional"/>
+		<xsd:attribute name="CC" type="xsd:string" use="optional">
+			<xsd:annotation>
+				<xsd:documentation>Confidence level of each character in that string. A list of numbers, one number between 0 (sure) and 9 (unsure) for each character.</xsd:documentation>
+			</xsd:annotation>
+		</xsd:attribute>
+		<xsd:attribute name="CS" type="xsd:boolean" use="optional">
+			<xsd:annotation>
+				<xsd:documentation>Correction Status. Indicates whether manual correction has been done or not. The correction status should be recorded at the highest level possible (Block, TextLine, String).</xsd:documentation>
+			</xsd:annotation>
+		</xsd:attribute>
+		<xsd:attribute name="LANG" type="xsd:language" use="optional">
+			<xsd:annotation>
+				<xsd:documentation>Attribute to record language of the string. The language should be recorded at the highest level possible.</xsd:documentation>
+			</xsd:annotation>
+		</xsd:attribute>
+	</xsd:complexType>
+	<xsd:complexType name="PageSpaceType">
+		<xsd:annotation>
+			<xsd:documentation>A region on a page</xsd:documentation>
+		</xsd:annotation>
+		<xsd:sequence>
+			<xsd:element name="Shape" type="ShapeType" minOccurs="0" maxOccurs="1"/>
+			<xsd:sequence minOccurs="0" maxOccurs="unbounded">
+				<xsd:group ref="BlockGroup"/>
+			</xsd:sequence>
+		</xsd:sequence>
+		<xsd:attribute name="ID" type="PageSpaceTypeID" use="optional"/>
+		<xsd:attribute name="STYLEREFS" type="xsd:IDREFS" use="optional"/>
+		<xsd:attribute name="PROCESSINGREFS" type="xsd:IDREFS" use="optional"/>
+		<xsd:attribute name="HEIGHT" type="xsd:float" use="optional"/>
+		<xsd:attribute name="WIDTH" type="xsd:float" use="optional"/>
+		<xsd:attribute name="HPOS" type="xsd:float" use="optional"/>
+		<xsd:attribute name="VPOS" type="xsd:float" use="optional"/>
+	</xsd:complexType>
+	<xsd:simpleType name="PointsType">
+		<xsd:annotation>
+			<xsd:documentation>A list of points</xsd:documentation>
+		</xsd:annotation>
+		<xsd:restriction base="xsd:string"/>
+	</xsd:simpleType>
+	<xsd:complexType name="ShapeType">
+		<xsd:annotation>
+			<xsd:documentation>Describes the bounding shape of a block, if it is not rectangular.</xsd:documentation>
+		</xsd:annotation>
+		<xsd:choice>
+			<xsd:element name="Polygon" type="PolygonType"/>
+			<xsd:element name="Ellipse" type="EllipseType"/>
+			<xsd:element name="Circle" type="CircleType"/>
+		</xsd:choice>
+	</xsd:complexType>
+	<xsd:simpleType name="InlineDirType">
+		<xsd:annotation>
+			<xsd:documentation>Describes the base direction of text inside a line or of all lines inside a text block.</xsd:documentation>
+			<xsd:documentation>These values should correspond to the base direction assigned to the respective elements during Unicode encoding with the BiDi algorithm.</xsd:documentation>

```

Comment:
```suggestion
			<xsd:documentation>
				Describes the base direction (from character to character) of text inside one line or all lines of a text block.
				These values should correspond to the base direction assigned to the respective elements during Unicode encoding with the BiDi algorithm.
			</xsd:documentation>
```

The second sentence is still not clear to me. The BiDi algorithm will assign directional states between characters to display them in the right order, right? Then what does `base` denote here, and why do you call this `encoding`?

Reply:
The BiDi algorithm has a base direction as a free parameter. This for example determines (among other things like mirroring) which direction weakly directional characters (sentence marks, parentheses, ...) are assigned outside of disambiguating context and the order of individual runs of different directionality. Decoding with the wrong base direction, i.e. a different one than used for inputting the original data, will flip runs around, place weakly directional characters at the wrong and of the run, and incorrectly mirror parentheses.

Systems that have their principal language set to something RTL usually set this base direction to RTL and vice versa. OCR engines encode the labels produced in display order into the logical order with a base directionality set for the BiDi algorithm as well. `Unicode encoding` in this context just means whatever process produced the Unicode text in its logical order and its respective settings for the BiDi algorithm.

There is nice page with examples about that whole issue at the [W3C](https://www.w3.org/International/articles/inline-bidi-markup/uba-basics#context).

Reply:
Thanks, I think I do understand now. (Still I think the `documentation` formulation should be improved for maximum clarity.)

But then IIUC the combination of 
- block and line-level base direction markup with
- BiDi algorithm for the text content

will not always be enough for compositionality anyway: As the more complicated cases in the W3C articles ("beyond bidi algorithm" and "inline bidi markup") show, in general you need directionality markup on arbitrarily recursive phrases/spans. (For example, mixed LTR/RTL content in quotations, titles, enumerations including punctuation.) However, ALTO's structural hierarchy is flat 2-level.

So I guess practically the text content may need to include Unicode control sequences (RLI+LRI+PDI, RLE+LRE+PDF, RLM+LRM) for that, which in turn will interact with the directionality markup.

Would you agree so far?

What is still unclear to me is why the value range of the new `@BASEDIRECTION` markup includes two values (`ttb` and `btt`) for vertical writing at all. IIUC the BiDi algorithm has nothing to say on horizontal vs vertical (since these cannot be mixed) and nothing to do in vertical (since no within-line changes can occur). So these two additional values are only relevant for redisplay purposes, or conversely, to OCR processing (i.e. how to flip/rotate the cropped line image for input into the LSTM) – right? 

However, how would we know whether a top-to-bottom line needs to be rotated 90° (e.g. for a vertical Chinese model or Latin script model in Chinese writing) or -90° (e.g. Arabic script in Chinese writing) for OCR input?

Reply:
The explicit direction control sequences are already in-band so there's no need to consider them for the purpose of ALTO. The base direction interacts with it only in that it is needed *in addition* to it to determine the correct display order of a sequence as there's no way to determine the correct setting just from the code point stream.

`@BASEDIRECTION` encodes both the BiDi base direction and the correct line orientation for display, that is correct. The Arabic text would be `btt` (implying 'R' base direction), the Chinese one `ttb` (implying 'L' base direction). That would be sufficient to rotate either line into the correct plane. 

Again, there's a handy W3C document on all of [that](https://www.w3.org/TR/css-writing-modes-3/). This https://github.com/altoxml/schema/issues/12#issuecomment-113184844 comment that was already linked in my initial post also explains it again.

Reply:
I've linked the W3C doc directly in the doc string and elaborated a bit on the possible values of `@BASEDIRECTION`. I think that should resolve the issues.